File: apps/api/src/main/java/vnpt/vsp/module/tournament/entity/TournamentPolicy.java
package vnpt.vsp.module.tournament.entity;

import jakarta.persistence.*;
import java.time.Instant;
import java.util.UUID;

/**
 * TournamentPolicy entity — feature restriction policy for tournament rounds.
 *
 * Per PRD §8.12: "Tournament Mode may lock after round start."
 * Per Architecture §14 QG-8: "Basic Tournament Mode restrictions are
 * represented in config and UI."
 *
 * A null policy means no restrictions (casual/practice modes).
 * A non-null policy activates feature gates on the mobile app.
 */
@Entity
@Table(name = "tournament_policies")
public class TournamentPolicy {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "name", nullable = false)
    private String name;

    @Column(name = "wind_adjustment", nullable = false)
    private boolean windAdjustment = true;

    @Column(name = "plays_like", nullable = false)
    private boolean playsLike = true;

    @Column(name = "elevation", nullable = false)
    private boolean elevation = true;

    @Column(name = "club_recommendation", nullable = false)
    private boolean clubRecommendation = true;

    @Column(name = "contours", nullable = false)
    private boolean contours = true;

    @Column(name = "putting_help", nullable = false)
    private boolean puttingHelp = true;

    @Column(name = "ai_features", nullable = false)
    private boolean aiFeatures = true;

    @Column(name = "is_locked", nullable = false)
    private boolean isLocked = false;

    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @Column(name = "created_by", nullable = false)
    private Long createdBy;

    @Version
    @Column(name = "version", nullable = false)
    private int version = 1;

    @PrePersist
    protected void onCreate() {
        createdAt = Instant.now();
    }

    // ─── Feature query ────────────────────────────────────────────────────────

    /**
     * Returns true if the named feature flag is enabled.
     *
     * @param flagName one of: windAdjustmentEnabled, playsLikeEnabled,
     *     elevationEnabled, clubRecommendationEnabled, contoursEnabled,
     *     puttingHelpEnabled, aiFeaturesEnabled
     */
    public boolean isFeatureEnabled(String flagName) {
        return switch (flagName) {
            case "windAdjustmentEnabled" -> windAdjustment;
            case "playsLikeEnabled" -> playsLike;
            case "elevationEnabled" -> elevation;
            case "clubRecommendationEnabled" -> clubRecommendation;
            case "contoursEnabled" -> contours;
            case "puttingHelpEnabled" -> puttingHelp;
            case "aiFeaturesEnabled" -> aiFeatures;
            default -> true;
        };
    }

    // ─── Lock ────────────────────────────────────────────────────────────────

    /**
     * Locks this policy, preventing feature flag changes without TournamentDirector role.
     * Idempotent: no-op if already locked.
     */
    public void lock() {
        this.isLocked = true;
    }

    // ─── Getters / Setters ──────────────────────────────────────────────────

    public UUID getId() { return id; }
    public void setId(UUID id) { this.id = id; }

    public String getName() { return name; }
    public void setName(String name) { this.name = name; }

    public boolean isWindAdjustment() { return windAdjustment; }
    public void setWindAdjustment(boolean windAdjustment) { this.windAdjustment = windAdjustment; }

    public boolean isPlaysLike() { return playsLike; }
    public void setPlaysLike(boolean playsLike) { this.playsLike = playsLike; }

    public boolean isElevation() { return elevation; }
    public void setElevation(boolean elevation) { this.elevation = elevation; }

    public boolean isClubRecommendation() { return clubRecommendation; }
    public void setClubRecommendation(boolean clubRecommendation) { this.clubRecommendation = clubRecommendation; }

    public boolean isContours() { return contours; }
    public void setContours(boolean contours) { this.contours = contours; }

    public boolean isPuttingHelp() { return puttingHelp; }
    public void setPuttingHelp(boolean puttingHelp) { this.puttingHelp = puttingHelp; }

    public boolean isAiFeatures() { return aiFeatures; }
    public void setAiFeatures(boolean aiFeatures) { this.aiFeatures = aiFeatures; }

    public boolean isLocked() { return isLocked; }
    public void setLocked(boolean locked) { isLocked = locked; }

    public Instant getCreatedAt() { return createdAt; }
    public void setCreatedAt(Instant createdAt) { this.createdAt = createdAt; }

    public Long getCreatedBy() { return createdBy; }
    public void setCreatedBy(Long createdBy) { this.createdBy = createdBy; }

    public int getVersion() { return version; }
    public void setVersion(int version) { this.version = version; }
}

File: apps/mobile/lib/domain/models/tournament_policy.dart
// TournamentPolicy Model — VSP Mobile App
//
// Domain model for a tournament's feature restriction policy.
//
// Per PRD §8.12: "Tournament Mode may lock after round start."
// Per Architecture §14 QG-8: "Basic Tournament Mode restrictions are
// represented in config and UI."
//
// Story 7.4 — Slice A: TournamentPolicy Domain Model

import 'package:equatable/equatable.dart';

import 'tournament_feature.dart';

/// Result of attempting to modify a locked policy.
class TournamentPolicyLockedException implements Exception {
  final String message;
  const TournamentPolicyLockedException([
    this.message = 'Policy is locked and cannot be modified',
  ]);

  @override
  String toString() => 'TournamentPolicyLockedException: $message';
}

/// Domain model for a tournament feature restriction policy.
///
/// A policy is attached to a tournament-format round to control which
/// assistance features are available to the golfer.
///
/// **Null policy** (no policy attached): All features enabled — casual/practice modes.
///
/// **Non-null policy**: Only features with `enabled = true` are available.
/// All other features must be hidden or disabled with an explanation tooltip.
///
/// When `isLocked == true`, no feature flag may be changed without the
/// `TournamentDirector` role.
class TournamentPolicy extends Equatable {
  /// Unique identifier for this policy.
  final String id;

  /// Human-readable name for this policy.
  final String name;

  /// Whether wind adjustment is enabled in this tournament.
  final bool windAdjustmentEnabled;

  /// Whether plays-like distance is enabled in this tournament.
  final bool playsLikeEnabled;

  /// Whether elevation data is enabled in this tournament.
  final bool elevationEnabled;

  /// Whether club recommendation is enabled in this tournament.
  final bool clubRecommendationEnabled;

  /// Whether green contours are enabled in this tournament.
  final bool contoursEnabled;

  /// Whether putting help is enabled in this tournament.
  final bool puttingHelpEnabled;

  /// Whether AI features are enabled in this tournament.
  final bool aiFeaturesEnabled;

  /// When true, feature flags cannot be changed without TournamentDirector role.
  final bool isLocked;

  /// Timestamp when this policy was created.
  final DateTime createdAt;

  /// ID of the user or system that created this policy.
  final String createdBy;

  /// Version number for optimistic locking.
  final int version;

  const TournamentPolicy({
    required this.id,
    required this.name,
    this.windAdjustmentEnabled = true,
    this.playsLikeEnabled = true,
    this.elevationEnabled = true,
    this.clubRecommendationEnabled = true,
    this.contoursEnabled = true,
    this.puttingHelpEnabled = true,
    this.aiFeaturesEnabled = true,
    this.isLocked = false,
    required this.createdAt,
    required this.createdBy,
    this.version = 1,
  });

  /// Returns true if the given [feature] is enabled in this policy.
  ///
  /// Convenience method that maps [TournamentFeature] to the correct flag.
  bool isFeatureEnabled(TournamentFeature feature) {
    switch (feature) {
      case TournamentFeature.windAdjustment:
        return windAdjustmentEnabled;
      case TournamentFeature.playsLike:
        return playsLikeEnabled;
      case TournamentFeature.elevation:
        return elevationEnabled;
      case TournamentFeature.clubRecommendation:
        return clubRecommendationEnabled;
      case TournamentFeature.contours:
        return contoursEnabled;
      case TournamentFeature.puttingHelp:
        return puttingHelpEnabled;
      case TournamentFeature.aiFeatures:
        return aiFeaturesEnabled;
    }
  }

  /// Returns the reason why [feature] is restricted, for display in UI.
  ///
  /// Returns null if the feature is enabled (no restriction).
  String? getRestrictionReason(TournamentFeature feature) {
    if (isFeatureEnabled(feature)) return null;
    return feature.restrictionLabel;
  }

  /// Locks this policy, preventing further feature flag changes.
  ///
  /// Idempotent: if already locked, returns this unchanged.
  /// Throws [TournamentPolicyLockedException] if called on an already-locked
  /// policy from a non-TournamentDirector caller.
  TournamentPolicy lock() {
    return copyWith(isLocked: true);
  }

  /// Returns a copy of this policy with [feature] set to [enabled].
  ///
  /// Throws [TournamentPolicyLockedException] if this policy is locked
  /// and [actorHasDirectorRole] is false.
  TournamentPolicy setFeature({
    required TournamentFeature feature,
    required bool enabled,
    required bool actorHasDirectorRole,
  }) {
    if (isLocked && !actorHasDirectorRole) {
      throw const TournamentPolicyLockedException();
    }

    switch (feature) {
      case TournamentFeature.windAdjustment:
        return copyWith(windAdjustmentEnabled: enabled, version: version + 1);
      case TournamentFeature.playsLike:
        return copyWith(playsLikeEnabled: enabled, version: version + 1);
      case TournamentFeature.elevation:
        return copyWith(elevationEnabled: enabled, version: version + 1);
      case TournamentFeature.clubRecommendation:
        return copyWith(
          clubRecommendationEnabled: enabled,
          version: version + 1,
        );
      case TournamentFeature.contours:
        return copyWith(contoursEnabled: enabled, version: version + 1);
      case TournamentFeature.puttingHelp:
        return copyWith(puttingHelpEnabled: enabled, version: version + 1);
      case TournamentFeature.aiFeatures:
        return copyWith(aiFeaturesEnabled: enabled, version: version + 1);
    }
  }

  // ─── JSON ─────────────────────────────────────────────────────────────────

  factory TournamentPolicy.fromJson(Map<String, dynamic> json) {
    return TournamentPolicy(
      id: json['id'] as String,
      name: json['name'] as String,
      windAdjustmentEnabled: json['windAdjustmentEnabled'] as bool? ?? true,
      playsLikeEnabled: json['playsLikeEnabled'] as bool? ?? true,
      elevationEnabled: json['elevationEnabled'] as bool? ?? true,
      clubRecommendationEnabled:
          json['clubRecommendationEnabled'] as bool? ?? true,
      contoursEnabled: json['contoursEnabled'] as bool? ?? true,
      puttingHelpEnabled: json['puttingHelpEnabled'] as bool? ?? true,
      aiFeaturesEnabled: json['aiFeaturesEnabled'] as bool? ?? true,
      isLocked: json['isLocked'] as bool? ?? false,
      createdAt: DateTime.parse(json['createdAt'] as String),
      createdBy: json['createdBy'] as String,
      version: (json['version'] as num?)?.toInt() ?? 1,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'name': name,
    'windAdjustmentEnabled': windAdjustmentEnabled,
    'playsLikeEnabled': playsLikeEnabled,
    'elevationEnabled': elevationEnabled,
    'clubRecommendationEnabled': clubRecommendationEnabled,
    'contoursEnabled': contoursEnabled,
    'puttingHelpEnabled': puttingHelpEnabled,
    'aiFeaturesEnabled': aiFeaturesEnabled,
    'isLocked': isLocked,
    'createdAt': createdAt.toIso8601String(),
    'createdBy': createdBy,
    'version': version,
  };

  // ─── Copy ─────────────────────────────────────────────────────────────────

  TournamentPolicy copyWith({
    String? id,
    String? name,
    bool? windAdjustmentEnabled,
    bool? playsLikeEnabled,
    bool? elevationEnabled,
    bool? clubRecommendationEnabled,
    bool? contoursEnabled,
    bool? puttingHelpEnabled,
    bool? aiFeaturesEnabled,
    bool? isLocked,
    DateTime? createdAt,
    String? createdBy,
    int? version,
  }) {
    return TournamentPolicy(
      id: id ?? this.id,
      name: name ?? this.name,
      windAdjustmentEnabled:
          windAdjustmentEnabled ?? this.windAdjustmentEnabled,
      playsLikeEnabled: playsLikeEnabled ?? this.playsLikeEnabled,
      elevationEnabled: elevationEnabled ?? this.elevationEnabled,
      clubRecommendationEnabled:
          clubRecommendationEnabled ?? this.clubRecommendationEnabled,
      contoursEnabled: contoursEnabled ?? this.contoursEnabled,
      puttingHelpEnabled: puttingHelpEnabled ?? this.puttingHelpEnabled,
      aiFeaturesEnabled: aiFeaturesEnabled ?? this.aiFeaturesEnabled,
      isLocked: isLocked ?? this.isLocked,
      createdAt: createdAt ?? this.createdAt,
      createdBy: createdBy ?? this.createdBy,
      version: version ?? this.version,
    );
  }

  @override
  List<Object?> get props => [
    id,
    name,
    windAdjustmentEnabled,
    playsLikeEnabled,
    elevationEnabled,
    clubRecommendationEnabled,
    contoursEnabled,
    puttingHelpEnabled,
    aiFeaturesEnabled,
    isLocked,
    createdAt,
    createdBy,
    version,
  ];
}
